charms/kfp-persistence/src/charm.py

```diff
@@ -96,14 +81,109 @@ def set_pod_spec(self, event):
+            raise CheckFailedError(str(err), BlockedStatus)
+        return interfaces
+
+    def _get_kfpapi(self, interfaces):
+        relation_name = "kfp-api"
+        return self.validate_sdi_interface(interfaces, relation_name)
+
+    def validate_sdi_interface(self, interfaces: dict, relation_name: str, default_return=None):
+        """Validates data received from SerializedDataInterface, returning the data if valid.
+
+        Optionally can return a default_return value when no relation is established
+
+        Raises:
+            CheckFailed(..., Blocked) when no relation established (unless default_return set)
+            CheckFailed(..., Blocked) if interface is not using SDI
+            CheckFailed(..., Blocked) if data in interface fails schema check
+            CheckFailed(..., Waiting) if we have a relation established but no data passed
+
+        Params:
+            interfaces:
+
+        Returns:
+              (dict) interface data
+        """
+        # If nothing is related to this relation, return a default value or raise an error
+        if relation_name not in interfaces or interfaces[relation_name] is None:
+            if default_return is not None:
+                return default_return
+            else:
+                raise CheckFailedError(
+                    f"Missing required relation for {relation_name}", BlockedStatus
+                )
+
+        relations = interfaces[relation_name]
+        if not isinstance(relations, SerializedDataInterface):
+            raise CheckFailedError(
+                f"Unexpected error with {relation_name} relation data - data not as expected",
+                BlockedStatus,
+            )
+
+        # Get and validate data from the relation
+        try:
+            # relations is a dict of {(ops.model.Relation, ops.model.Application): data}
+            unpacked_relation_data = relations.get_data()
+        except ValidationError as val_error:
+            # Validation in .get_data() ensures if data is populated, it matches the schema and is
+            # not incomplete
+            self.log.error(val_error)
+            raise CheckFailedError(
+                f"Found incomplete/incorrect relation data for {relation_name}.  See logs",
+                BlockedStatus,
+            )
+
+        # Check if we have an established relation with no data exchanged
+        if len(unpacked_relation_data) == 0:
+            raise CheckFailedError(f"Waiting for {relation_name} relation data", WaitingStatus)
+
+        # Unpack data (we care only about the first element)
+        data_dict = list(unpacked_relation_data.values())[0]
+
+        # Catch if empty data dict is received (JSONSchema ValidationError above does not raise
+        # when this happens)
+        if len(data_dict) == 0:
+            raise CheckFailedError(
+                f"Found incomplete/incorrect relation data for {relation_name}.",
+                BlockedStatus,
+            )
```

Comment:
Same as the other [review](https://github.com/canonical/kfp-operators/pull/27#discussion_r759448470)